SAT/Z3Interfacing.hpp

```diff
@@ -141,19 +141,55 @@ class Z3Interfacing : public PrimitiveProofRecordingSATSolver
 
   struct FuncOrPredId
   {
-    explicit FuncOrPredId(unsigned id, bool isPredicate) : id(id), isPredicate(isPredicate) {}
-    explicit FuncOrPredId(Term* term) : FuncOrPredId(term->functor(), term->isLiteral()) {}
-    static FuncOrPredId function(FuncId id) { return FuncOrPredId ( id, false ); }
-    static FuncOrPredId predicate(PredId id) { return FuncOrPredId ( id, true ); }
+    explicit FuncOrPredId(unsigned id, bool isPredicate, Term *forSorts = nullptr) : id(id), isPredicate(isPredicate), forSorts(forSorts) {}
+    explicit FuncOrPredId(Term* term) :
+      FuncOrPredId(
+        term->functor(),
+        term->isLiteral(),
+        term->numTypeArguments() == 0
+          ? nullptr
+          : term
```

Comment:
Another option, would be to always set `forSorts` in both the mono and the poly case. This would lead to some minor code simplifications below.

Reply:
Sadly there are some places where we don't (?) have access to a concrete term. I guess in principle we could create a "dummy" term or review all of these places if you are very keen?

Reply:
Sorry, I don't follow what you mean by "concrete term". We are passing a term into the constructor, can we not just use that?

Reply:
The term is defaulted to `nullptr` in the first constructor (i.e. not this one) because at the call site we have access to a symbol, but not a term using that symbol. So `forSorts` might be null later regardless of what we do here - but you're right, for this particular bit of code we can just use the term.